File: octopoints-kotlin/app/src/main/java/dev/tpavani/octopoints/service/impl/DBTeamService.kt
package dev.tpavani.octopoints.service.impl

import dev.tpavani.octopoints.db.DbSingleton
import dev.tpavani.octopoints.db.dao.JoinTeamDao
import dev.tpavani.octopoints.db.dao.TeamDao
import dev.tpavani.octopoints.db.entity.JoinTeamEntity
import dev.tpavani.octopoints.db.entity.TeamEntity
import dev.tpavani.octopoints.service.DBService
import dev.tpavani.octopoints.service.model.Team
import dev.tpavani.octopoints.service.model.User
import dev.tpavani.octopoints.service.utils.GetListFilter
import java.util.function.Predicate

class DBTeamService: DBService<Team, TeamEntity>() {

    override val dao: TeamDao = DbSingleton.getDb().teamDao()
    private val joinTeamDao: JoinTeamDao = DbSingleton.getDb().joinTeamDao()

    override fun entityMapper(model: Team): TeamEntity {
        return TeamEntity(model.getId(), model.getMatchId(), model.getName(), model.total, model.partial)
    }

    override suspend fun getList(filter: GetListFilter): MutableList<Team> {
        if(filter.id != null){
            val list = dao.getTeamsByMatchId(filter.id!!).map { Team(it.id, it.matchId, it.name, it.total, it.partial) }.toMutableList()
            val iterator: Iterator<Team> = list.iterator()
            while (iterator.hasNext()){
                var team = iterator.next()
                team.users = joinTeamDao.getTeammates(team.getId()).map { User(it.id, it.username, it.win, it.lose) }.toMutableList()
            }
            return list
        }
        return mutableListOf()
    }

    suspend fun joinTeam(team: Team, user: User) {
        joinTeamDao.create(JoinTeamEntity(team.getId(), user.getId()))
        team.users.add(user)
    }

    suspend fun leaveTeam(team: Team, user: User) {
        joinTeamDao.remove(JoinTeamEntity(team.getId(), user.getId()))
        team.users.remove(user)
    }
}

File: octopoints_flutter/lib/service/impl/DBTeamService.dart
import 'package:octopoints_flutter/db/DbSingleton.dart';
import 'package:octopoints_flutter/db/dao/JoinTeamDao.dart';
import 'package:octopoints_flutter/db/dao/TeamDao.dart';
import 'package:octopoints_flutter/db/entity/JoinTeamEntity.dart';
import 'package:octopoints_flutter/db/entity/TeamEntity.dart';
import 'package:octopoints_flutter/db/entity/UserEntity.dart';
import 'package:octopoints_flutter/service/DBService.dart';
import 'package:octopoints_flutter/service/model/Team.dart';
import 'package:octopoints_flutter/service/model/User.dart';
import 'package:octopoints_flutter/service/utils/GetListFilter.dart';

class DBTeamService extends DBService<Team, TeamEntity> {
  @override
  TeamDao get dao => DbSingleton().db.teamDao;

  final JoinTeamDao _joinTeamDao = DbSingleton().db.joinTeamDao;

  @override
  TeamEntity entityMapper(Team model) {
    return TeamEntity(model.id == 0 ? null : model.id, model.matchId, model.name, model.total, model.partial);
  }

  @override
  Future<List<Team>> getList(GetListFilter filters) async {
    if(filters.id != null){
      List<TeamEntity> list = await dao.getTeamsByMatchId(filters.id!);
      List<Team> teams = list.map((e) => Team(e.matchId, e.name,id: e.id!, total: e.total, partial: e.partial)).toList();
      Iterator iterator = teams.iterator;
      while(iterator.moveNext()){
        Team team = iterator.current;
        List<UserEntity> users = await _joinTeamDao.getTeammates(team.id);
        team.users = users.map((e) => User(e.username, win: e.win, lose: e.lose, id: e.id!)).toList();
      }
      return teams;
    }
    return [];
  }

  Future<void> joinTeam(Team team, User user) async {
    await _joinTeamDao.create(JoinTeamEntity(team.id, user.id));
    team.users.add(user);
  }

  Future<void> leaveTeam(Team team, User user) async {
    await _joinTeamDao.remove(JoinTeamEntity(team.id, user.id));
    team.users.remove(user);
  }

}